cmake/modules/CMakeLists.txt

```diff
@@ -1,19 +1,58 @@
 set(swift_cmake_builddir "${SWIFT_BINARY_DIR}/${SWIFT_INSTALL_PACKAGE_DIR}")
 
-# Generate build-tree exports list only
-set(SWIFT_EXPORTS_FILE ${swift_cmake_builddir}/SwiftExports.cmake)
+set(SWIFT_EXPORTS_FILE "${swift_cmake_builddir}/SwiftExports.cmake")
 get_property(SWIFT_EXPORTS GLOBAL PROPERTY SWIFT_EXPORTS)
-get_property(SWIFT_BUILDTREE_EXPORTS GLOBAL PROPERTY SWIFT_BUILDTREE_EXPORTS)
-
-set(SWIFT_CONFIG_EXPORTS ${SWIFT_EXPORTS} ${SWIFT_BUILDTREE_EXPORTS})
-export(TARGETS ${SWIFT_CONFIG_EXPORTS} FILE ${SWIFT_EXPORTS_FILE})
+export(TARGETS ${SWIFT_EXPORTS} FILE ${SWIFT_EXPORTS_FILE})
 
+set(SWIFT_BUILDTREE_EXPORTS_FILE "${swift_cmake_builddir}/SwiftBuildTreeExports.cmake")
+get_property(SWIFT_BUILDTREE_EXPORTS GLOBAL PROPERTY SWIFT_BUILDTREE_EXPORTS)
+export(TARGETS ${SWIFT_BUILDTREE_EXPORTS} FILE ${SWIFT_BUILDTREE_EXPORTS_FILE})
 
-set(SWIFT_INCLUDE_DIRS ${SWIFT_INCLUDE_DIR} ${SWIFT_MAIN_INCLUDE_DIR})
-set(SWIFT_LIBRARY_DIRS ${SWIFT_LIBRARY_OUTPUT_INTDIR})
+# Generate SwiftConfig.cmake for the build tree
+set(SWIFT_CONFIG_EXPORTS "${SWIFT_EXPORTS};${SWIFT_BUILDTREE_EXPORTS}")
+set(SWIFT_CONFIG_EXPORTS_FILE "${SWIFT_EXPORTS_FILE}")
+set(SWIFT_CONFIG_BINARY_DIR "${SWIFT_BINARY_DIR}")
+set(SWIFT_CONFIG_INCLUDE_DIRS ${SWIFT_INCLUDE_DIR} ${SWIFT_MAIN_INCLUDE_DIR})
+set(SWIFT_CONFIG_LIBRARY_DIRS ${SWIFT_LIBRARY_OUTPUT_INTDIR})
+set(SWIFT_CONFIG_CMAKE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+set(swift_config_include_buildtree_exports
+"include(\"${SWIFT_BUILDTREE_EXPORTS_FILE}\")")
 
 configure_file(
   SwiftConfig.cmake.in
   ${swift_cmake_builddir}/SwiftConfig.cmake
-  @ONLY)
\ No newline at end of file
+  @ONLY)
+set(swift_config_include_buildtree_exports)
+
+# Generate SwiftConfig.cmake for the install tree
+set(SWIFT_CONFIG_CODE
+"# Compute the installation prefix from this SwiftConfig.cmake location.
+get_filename_component(SWIFT_INSTALL_PREFIX \"\${CMAKE_CURRENT_LIST_FILE}\" PATH)")
+# Construct the proper number of get_filename_component(... PATH)
+# calls to compute the installation prefix.
+string(REGEX REPLACE "/" ";" _count "${SWIFT_INSTALL_PACKAGE_DIR}")
+foreach(p ${_count})
+  set(SWIFT_CONFIG_CODE "${SWIFT_CONFIG_CODE}
+get_filename_component(SWIFT_INSTALL_PREFIX \"\${SWIFT_INSTALL_PREFIX}\" PATH)")
+endforeach(p)
+set(SWIFT_CONFIG_EXPORTS "${SWIFT_EXPORTS}")
+set(SWIFT_CONFIG_EXPORTS_FILE "\${SWIFT_CMAKE_DIR}/SwiftExports.cmake")
+set(SWIFT_CONFIG_BINARY_DIR "\${SWIFT_INSTALL_PREFIX}")
+set(SWIFT_CONFIG_INCLUDE_DIRS "\${SWIFT_INSTALL_PREFIX}/include")
+set(SWIFT_CONFIG_LIBRARY_DIRS "\${SWIFT_INSTALL_PREFIX}/lib\${LLVM_LIBDIR_SUFFIX}")
+set(SWIFT_CONFIG_CMAKE_DIR "\${SWIFT_INSTALL_PREFIX}/${SWIFT_INSTALL_PACKAGE_DIR}")
+configure_file(
+  SwiftConfig.cmake.in
+  ${CMAKE_CURRENT_BINARY_DIR}/CMakeFiles/SwiftConfig.cmake
+  @ONLY)
+
+if (NOT LLVM_INSTALL_TOOLCHAIN_ONLY)
+  get_property(swift_has_exports GLOBAL PROPERTY SWIFT_HAS_EXPORTS)
+  if(swift_has_exports)
+    install(EXPORT SwiftExports DESTINATION ${SWIFT_INSTALL_PACKAGE_DIR})
```

Comment:
I need to look at how LLVM does this before I can give a +1. That is why I was wondering if it is something we can refactor into LLVM and just invoke here.

Reply:
This is essentially what LLVM and Clang do. It is probably something we can refactor into LLVM but it's about 3-4 lines. Not sure if the layer of indirection is worth the trouble here.